### ISSUE TITLE
Datetime not parse local format 4.0

### ISSUE THREAD

**Issue by mcsknp** [cakephp/cakephp#14071]
This is a (multiple allowed):

* [x] bug
* [ ] enhancement
* [x] feature-discussion (RFC)

* CakePHP Version: 4.0
* Platform and Target: Mac OS, Nginx

### What you did
Set datetime field from `dd-MM-yyyy HH:mm` format

### What happened
Hour and minute is not parsed when I use below code
```
TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat([\IntlDateFormatter::SHORT, -1]);
```
It will be null if I use below code
```
TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat('dd-MM-yyyy HH:mm:ss');
```
There are no problem when create date with below code
```
new \Cake\I18n\FrozenTime($dateString)
```

### What you expected to happen
Timestamp is parsed

**Comment by othercorey** [comment]
Can you give us an example date string in your form?

**Comment by mcsknp** [comment]
I tried with `24-12-2019 10:00`

**Comment by othercorey** [comment]
> Hour and minute is not parsed when I use below code
> 
> ```
> TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat([\IntlDateFormatter::SHORT, -1]);
> ```

This is invalid for both date and time.  You must set both in array. I was able to parse with `dd-MM-yyyy HH:mm` format.

**Comment by mcsknp** [comment]
I tried with `dd-MM-yyyy HH:mm` but it can not parse second part of `dd-MM-yyyy HH:mm:ss`

**Comment by othercorey** [comment]
I don't understand.  Can you post the code you're using?

**Comment by mcsknp** [comment]
I want to parse both of `dd-MM-yyyy HH:mm` and `dd-MM-yyyy HH:mm:ss` like constructor of FrozenTime class.
I tried many pattern with below code
```
\Cake\Database\TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat([\IntlDateFormatter::SHORT, -1]);
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59'));
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59:35'));
\Cake\Database\TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat([\IntlDateFormatter::SHORT, \IntlDateFormatter::SHORT]);
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59'));
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59:35'));
\Cake\Database\TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat('dd-MM-yyyy HH:mm');
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59'));
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59:35'));
\Cake\Database\TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat('dd-MM-yyyy HH:mm:ss');
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59'));
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59:35'));
```
and result is
```
object(Cake\I18n\FrozenTime) {
	'time' => '2019-12-24 00:00:00.000000+09:00',
	'timezone' => 'Asia/Tokyo',
	'fixedNowTime' => false
}

object(Cake\I18n\FrozenTime) {
	'time' => '2019-12-24 00:00:00.000000+09:00',
	'timezone' => 'Asia/Tokyo',
	'fixedNowTime' => false
}

null

null

object(Cake\I18n\FrozenTime) {
	'time' => '2019-12-24 23:59:00.000000+09:00',
	'timezone' => 'Asia/Tokyo',
	'fixedNowTime' => false
}

object(Cake\I18n\FrozenTime) {
	'time' => '2019-12-24 23:59:00.000000+09:00',
	'timezone' => 'Asia/Tokyo',
	'fixedNowTime' => false
}

null

object(Cake\I18n\FrozenTime) {
	'time' => '2019-12-24 23:59:35.000000+09:00',
	'timezone' => 'Asia/Tokyo',
	'fixedNowTime' => false
}
```

And if it parse by `new FrozenTime`
```
debug(new \Cake\I18n\FrozenTime('24-12-2019 23:59'));
debug(new \Cake\I18n\FrozenTime('24-12-2019 23:59:35'));
```
result is
```
object(Cake\I18n\FrozenTime) {
	'time' => '2019-12-24 23:59:00.000000+09:00',
	'timezone' => 'Asia/Tokyo',
	'fixedNowTime' => false
}

object(Cake\I18n\FrozenTime) {
	'time' => '2019-12-24 23:59:35.000000+09:00',
	'timezone' => 'Asia/Tokyo',
	'fixedNowTime' => false
}
```

**Comment by mcsknp** [comment]
I found below code in `marshal` method of `\Cake\Database\Type\DatetimeType`
```
if (ctype_digit($value)) {
    /** @var \DateTimeInterface */
    return new $class('@' . $value);
} elseif ($isString && $this->_useLocaleMarshal) {
    return $this->_parseLocaleValue($value);
} elseif ($isString) {
    return $this->_parseValue($value);
}
```
Why locale and other string are separated while FrozenTime class is support both of them in constructor?
What do you thing if it will be
```
if (ctype_digit($value)) {
    /** @var \DateTimeInterface */
    return new $class('@' . $value);
} elseif ($isString) {
    /** @var \DateTimeInterface */
    return new $class($value);
}
```

**Comment by othercorey** [comment]
I understand now.  Duplicate of https://github.com/cakephp/cakephp/issues/14028.

You could disable the locale parser if the FrozenTime constructor works for you.

**Comment by markstory** [comment]
> Why locale and other string are separated while FrozenTime class is support both of them in constructor?

Because whilst the constructor supports *your* local datetime format. It might not support all locale based datetime formats.

The example of 
```php
\Cake\Database\TypeFactory::build('datetime')->useLocaleParser()->setLocaleFormat('dd-MM-yyyy HH:mm:ss');
debug(\Cake\Database\TypeFactory::build('datetime')->marshal('24-12-2019 23:59'));
```
Should return null to me. Your value is lacking seconds which is required in the datetime format string. However, the Intl constant variations should work and we can get those working again.

To help reproduce this problem what is your default locale set to?

**Comment by mcsknp** [comment]
I'm testing with `vi_VN`

**Comment by markstory** [comment]
I took a look at this and you are using an incompatible date/time format based on what libicu is expecting for the Vietnamese locale. The [icu rules](http://demo.icu-project.org/icu-bin/locexp?d_=en&_=vi_VN) indicate that the accepted Date-Time Pattern is required to be `time, date`. In local testing I was able to get the desired results using the following:

```php
date_default_timezone_set('Asia/Tokyo');
I18n::setLocale('vi_VN');
$this->type->useLocaleParser()->setLocaleFormat([\IntlDateFormatter::SHORT, \IntlDateFormatter::SHORT]);
debug($this->type->marshal('03:31, 29/12/2019'));
```

This example uses the format outlined in the icu documentation. It sounds like you want to use a format that doesn't exactly match the icu features, so you'll have to use a custom pattern.

I'm going to close this as if your input values match the format icu expects then locale parsing wants. Whether or not icu is using the most correct format for the Vietnamese locale is well outside of my control unfortunately.
